### PR TITLE
Secrets should be passed to pods using volumes instead of API calls

### PR DESCRIPTION
**Pull request by sarabala1979**
Secrets should be passed to pods using volumes instead of API calls

### CODE REVIEW

**Comment by xianlubird** [comment]
Awesome pr.  

LGTM

File: workflow/executor/executor.go [hunk 1]
```diff
@@ -421,19 +421,42 @@ func (we *WorkflowExecutor) SaveLogs() (*wfv1.Artifact, error) {
 	return &art, nil
 }
 
+// Retrive the Secrets from VolumeMount
```
Comment (jessesuen) [inline workflow/executor/executor.go]: Function comments should use name of function in sentence. e.g.:
```
// GetSecretFromVolMount retrieves the secrets from the volume mount
```
https://blog.golang.org/godoc-documenting-go-code

File: workflow/controller/workflowpod.go [hunk 2]
```diff
@@ -747,3 +757,97 @@ func verifyResolvedVariables(obj interface{}) error {
 	})
 	return unresolvedErr
 }
+
+// Retrieve and create Volumes and Volumemount object for Pod
```
Comment (jessesuen) [inline workflow/controller/workflowpod.go]: Function comment should use name of function in a sentence

File: workflow/controller/workflowpod.go [hunk 3]
```diff
@@ -747,3 +757,97 @@ func verifyResolvedVariables(obj interface{}) error {
...
+	woc.createArgoArtifactsRepoSecret(tmpl, allVolumesMap, uniqueKeyMap)
+
+	for _, art := range tmpl.Outputs.Artifacts {
+		woc.createSecretVolume(allVolumesMap, art, uniqueKeyMap)
+	}
+	for _, art := range tmpl.Inputs.Artifacts {
+		woc.createSecretVolume(allVolumesMap, art, uniqueKeyMap)
+	}
+
+	if len(allVolumesMap) > 0 {
```
Comment (jessesuen) [inline workflow/controller/workflowpod.go]: The check for `len(allVolumesMap) > 0` is unnecessary. Golang can append to nil.

File: workflow/controller/workflowpod.go [hunk 4]
```diff
@@ -747,3 +757,97 @@ func verifyResolvedVariables(obj interface{}) error {
 	})
 	return unresolvedErr
 }
+
+// Retrieve and create Volumes and Volumemount object for Pod
+func (woc *wfOperationCtx) createSecretVolumes(tmpl *wfv1.Template) (secretVolumes []apiv1.Volume, secretVolMounts []apiv1.VolumeMount) {
```
Comment (jessesuen) [inline workflow/controller/workflowpod.go]: Style nit: the team generally dislikes using named return values.

File: workflow/controller/workflowpod.go [hunk 5]
```diff
@@ -128,6 +128,14 @@ func (woc *wfOperationCtx) createWorkflowPod(nodeName string, mainCtr apiv1.Cont
 		pod.ObjectMeta.Labels[common.LabelKeyControllerInstanceID] = woc.controller.Config.InstanceID
 	}
 
+	err := woc.addArchiveLocation(pod, tmpl)
+	if err != nil {
+		return nil, err
+	}
+
+	volumes, volumeMounts := woc.createSecretVolumes(tmpl)
```
Comment (jessesuen) [inline workflow/controller/workflowpod.go]: Can we move secret volume addition into addVolumeReferences(), where the other volume manipulation currently happens? Then all volume manipulations can happen in the same place. Inside addVolumeReferences(), we can iterate initContainers and containers, and add the volume mounts if the container name is `init` or `wait`.

**Comment by jessesuen** [comment]
Also, I misused the terminology. The commit message should be:

`Secrets should be passed to pods using volumes instead of API calls`

File: workflow/executor/executor.go [hunk 6]
```diff
@@ -421,19 +421,42 @@ func (we *WorkflowExecutor) SaveLogs() (*wfv1.Artifact, error) {
 	return &art, nil
 }
 
+// GetSecretFromVolMount will retrive the Secrets from VolumeMount
+func (we *WorkflowExecutor) GetSecretFromVolMount(accessKeyName string, accessKey string) ([]byte, error) {
+
+	file, err := os.Open(common.SecretVolMountPath + "/" + accessKeyName + "/" + accessKey)
```
Comment (jessesuen) [inline workflow/executor/executor.go]: I think this entire method could be replaced with the following one-liner
```
ioutil.ReadFile(filepath.Join(common.SecretVolMountPath, accessKeyName, accessKey))
```

**Pull request merged**